src/controller/java/AndroidDeviceControllerWrapper.cpp

```diff
@@ -137,6 +151,62 @@ void AndroidDeviceControllerWrapper::OnPairingDeleted(CHIP_ERROR error)
     CallJavaMethod("onPairingDeleted", static_cast<jint>(error));
 }
 
+CHIP_ERROR AndroidDeviceControllerWrapper::GenerateNodeOperationalCertificate(const chip::PeerId & peerId, const chip::ByteSpan & csr,
+                                                                                int64_t serialNumber, uint8_t * certBuf,
+                                                                                uint32_t certBufSize, uint32_t & outCertLen)
+{
+    jmethodID method;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    err = FindMethod(GetEnvForCurrentThread(), mJavaObjectRef, "onOpCSRGenerationComplete", "([B)V", &method);
+    if (err != CHIP_NO_ERROR)
+   {
+      ChipLogError(Controller, "Error invoking onOpCSRGenerationComplete: %d", err);
+      return err;
+   }
+
+    // Initializing the KeyPair.
+     Initialize();
+
+    chip::X509CertRequestParams request = { serialNumber, mIssuerId,         mNow, mNow + mValidity, true, peerId.GetFabricId(),
```

Comment:
What is this API? Is that a sample commissioning logic? The OpCSR needs to come from the device and the device is the one who should have generated the key pair.

Reply:
So this is my understanding
The CHIPDeviceController (running as part of the commissioner) currently automatically sends the OpCSR request to the device as part of the pairing process. It also generates and fills in a Nonce as part of it.
On getting the OpCSR from the device, it calls the delegate to request the certificate signing.

The commissioner application provides a reference to OperationalCredentialsDelegate object when it initializes chip::DeviceCommissioner object and when the device sends the OpCSR back, the chip::DeviceCommissioner calls the  function  GenerateNodeOperationalCertificate  in the delegate. 

I intend to use this delegate to the callbacks to get the OpCSR.